### PR TITLE
Feat: Reset question using dispute callback + combine settle and reportPayouts

### PR DESCRIPTION
**Pull request by JonathanAmenechi**
## Overview

- Use `priceDisputed` callback on OO to reset questions upon dispute
- Combine `settle` and `reportPayouts` into a single function: `resolve`
- Generate questionID from hash of ancillary data
- authed `reset` failsafe function, to be used if the `priceDisputed` callback reverts or fails

## Description

This PR updates the flow of question resolution. 

Happy path
1. Question is initialized, prepared and a price is requested on the OO.
2. OO proposers converge on the correct result and post it to the OO
3. Anyone can call `resolve` which collects the result from the OO and resolves the underlying CTF market

Not so happy path
1. Question is initialized, prepared and a price is requested on the OO.
2. If there is an invalid proposal, the price is disputed. The OO then automatically calls the `priceDisputed` callback on the Adapter, which resets the question, sending out a new request to the OO.
3. If the callback reverts, an `auth`-ed user can call `reset`, which will send out a new request to the OO.
4. Either way, once a request is undisputed, anyone can call `resolve` which collects the result from the OO and resolves the underlying CTF market.

## Testing instructions

`yarn test`

## Types of changes

- [x] Refactor/enhancement 
- [ ] Bug fix/behavior correction 
- [x] New feature 
- [ ] Breaking change 
- [ ] Other, additional 

## Status

- [ ] Prefix PR title with `WIP` if necessary (changes not yet made).
- [x] Add tests to cover changes as needed.
- [ ] Update documentation/changelog as needed.
- [x] Verify all tests run correctly in CI and pass.
- [x] Ready for review/merge.

### CODE REVIEW

File: contracts/UmaConditionalTokensBinaryAdapter.sol [hunk 1]
```diff
@@ -461,18 +413,62 @@ contract UmaCtfAdapter is Auth, BulletinBoard, ReentrancyGuard {
...
 
-    /// @notice Gets resolution data for the question from the OO
-    /// @param questionData - The parameters of the question
-    function _getResolutionData(QuestionData storage questionData) internal view returns (int256) {
+    /// @notice Resolves the underlying CTF market
+    /// @param questionID   - The unique questionID of the question
+    /// @param questionData - The question data parameters
+    function _resolve(bytes32 questionID, QuestionData storage questionData) internal {
+        // Get the price from the OO
+        int256 price = optimisticOracle.settleAndGetPrice(
```
Comment (L-Kov) [inline contracts/UmaConditionalTokensBinaryAdapter.sol:421]: can someone settle the market in front of us? does that cause any issues with us calling this function?

File: contracts/UmaConditionalTokensBinaryAdapter.sol [hunk 2]
```diff
@@ -156,162 +156,136 @@ contract UmaCtfAdapter is Auth, BulletinBoard, ReentrancyGuard {
...
 
-        return _settle(questionID, questionData);
+        // Resolve the underlying market
+        return _resolve(questionID, questionData);
     }
 
-    /// @notice Retrieves the expected payout of a settled question
+    /// @notice Retrieves the expected payout array of the question
     /// @param questionID - The unique questionID of the question
     function getExpectedPayouts(bytes32 questionID) public view returns (uint256[] memory) {
```
Comment (L-Kov) [inline contracts/UmaConditionalTokensBinaryAdapter.sol:193]: is `getExpectedPayouts` even needed any longer?

File: contracts/UmaConditionalTokensBinaryAdapter.sol [hunk 3]
```diff
@@ -395,49 +368,28 @@ contract UmaCtfAdapter is Auth, BulletinBoard, ReentrancyGuard {
...
         optimisticOracle.setEventBased(UmaConstants.YesOrNoIdentifier, requestTimestamp, ancillaryData);
 
+        // Ensure that the dispute callback flag is set
+        optimisticOracle.setCallbacks(
+            UmaConstants.YesOrNoIdentifier,
+            requestTimestamp,
+            ancillaryData,
+            false, // DO NOT set callback on priceProposed
+            true, // DO set callback on priceDisputed
+            false // DO NOT set callback on priceSettled
```
Comment (L-Kov) [inline contracts/UmaConditionalTokensBinaryAdapter.sol:378]: Do we want to consider leveraging the callback on `priceSettled` to remove need to call `resolve`?

**Comment by L-Kov** [comment]
Also there is no need for ipfs in the deployment process now right (also really just doesn't work with ipfs anymore)? This will require some reworking of the listener I believe since it grabs the ipfs object based on questionID. We should look at the listener and figure out what we want to do about that (I think we can just go direct to strapi from the deployment portal).

File: contracts/UmaConditionalTokensBinaryAdapter.sol [hunk 4]
```diff
@@ -156,162 +156,136 @@ contract UmaCtfAdapter is Auth, BulletinBoard, ReentrancyGuard {
...
-
-        // Payouts: [YES, NO]
-        uint256[] memory payouts = new uint256[](2);
-
-        // Valid prices are 0, 0.5 and 1
-        require(
-            resolutionData == 0 || resolutionData == 0.5 ether || resolutionData == 1 ether,
-            AdapterErrors.InvalidData
-        );
+        // Fetches price from OO
```
[the remark this answers is not in the document]
Comment (JonathanAmenechi) [reply contracts/UmaConditionalTokensBinaryAdapter.sol:198]: That should be OK, we want to be able to see the expected payouts even if the question is paused

File: contracts/UmaConditionalTokensBinaryAdapter.sol (hunk 1 above)
Comment (JonathanAmenechi) [reply contracts/UmaConditionalTokensBinaryAdapter.sol:421]: Yes but that will not cause any issues for us. If a request is already settled, settleAndGetPrice will just return the resolvedPrice

File: contracts/UmaConditionalTokensBinaryAdapter.sol (hunk 2 above)
Comment (JonathanAmenechi) [reply contracts/UmaConditionalTokensBinaryAdapter.sol:193]: This is a fair question, I'm not really sure. It's a nice utility function but it's not necessary anymore

**Review by mshrieve — APPROVED** [review]
Looks good to me. My only lingering concern is the ability to front-run `initializeQuestion` with a zero reward, but this PR didn't affect that.

**Comment by JonathanAmenechi** [comment]
> Also there is no need for ipfs in the deployment process now right (also really just doesn't work with ipfs anymore)? This will require some reworking of the listener I believe since it grabs the ipfs object based on questionID. We should look at the listener and figure out what we want to do about that (I think we can just go direct to strapi from the deployment portal).

Yeah i think that's the best way around this, upload straight to strapi from deployment portal. The other approach is to not use callbacks and stick with the existing infra

**Comment by mshrieve** [comment]
> > Also there is no need for ipfs in the deployment process now right (also really just doesn't work with ipfs anymore)? This will require some reworking of the listener I believe since it grabs the ipfs object based on questionID. We should look at the listener and figure out what we want to do about that (I think we can just go direct to strapi from the deployment portal).
> 
> 
> 
> Yeah i think that's the best way around this, upload straight to strapi from deployment portal. The other approach is to not use callbacks and stick with the existing infra

A third option is to use the hash of the ancillary data as an internal identifier, in order to handle callbacks, and keep questionId externally provided, stored in the Question struct.

**Comment by JonathanAmenechi** [comment]
> > > Also there is no need for ipfs in the deployment process now right (also really just doesn't work with ipfs anymore)? This will require some reworking of the listener I believe since it grabs the ipfs object based on questionID. We should look at the listener and figure out what we want to do about that (I think we can just go direct to strapi from the deployment portal).
> > 
> > 
> > Yeah i think that's the best way around this, upload straight to strapi from deployment portal. The other approach is to not use callbacks and stick with the existing infra
> 
> A third option is to use the hash of the ancillary data as an internal identifier, in order to handle callbacks, and keep questionId externally provided, stored in the Question struct.

That could work but would mean we'd need to:

1) Store another mapping of questionID to internalID, to ensure that `resolve` and `reset` work as expected. So total of 2 mappings, `internalID` to `QuestionData` and `questionID` to `internalID`

2) Modify resolve and reset, so that they take the `ancillaryData` and `requestTimestamp` as arguments(imo, this isn't feasible)

**Comment by L-Kov** [comment]
>Yeah i think that's the best way around this, upload straight to strapi from deployment portal. The other approach is to not use callbacks and stick with the existing infra
I'd prefer going straight to strapi since we are already including too much information (extra info that really is only relevant to us) in ipfs object and pinata is making us pay now.

**Review by L-Kov — APPROVED** [review]

**Pull request merged**